src/shared/backend/graph/firmware/cloudbuild.ts

```diff
@@ -0,0 +1,192 @@
+      resolve: async (_, __, { cloudbuild, github }) => {
+        const [cloudTargets, githubReleases] = await Promise.all([
+          cloudbuild.fetchTargets(),
+          github("GET /repos/{owner}/{repo}/releases", {
+            owner: config.github.organization,
+            repo: config.github.repos.firmware,
+          }),
+        ]);
+        const firmwaresReleases = new Set(Object.keys(cloudTargets.releases));
+
+        // product of github releases and firmwares releases
+        const releases = githubReleases.data
+          .filter((release) => firmwaresReleases.has(release.tag_name))
+          .map((release) => {
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            const cloudRelease = cloudTargets.releases[release.tag_name]!;
+            return {
+              id: release.tag_name,
+              name: release.name ?? release.tag_name,
+              description: release.body_text,
+              isPrerelease: release.prerelease,
+              timestamp: release.published_at ?? release.created_at,
+              sha: cloudRelease.sha,
+              excludeTargets: cloudRelease.exclude_targets ?? [],
+            };
+          });
+
+        const targets = Object.entries(cloudTargets.targets).map(
+          ([id, target]) => ({
+            id,
+            name: target.description,
+            tags: target.tags ?? [],
+          })
+        );
+
+        const flags = Object.entries(cloudTargets.flags).map(([id, flag]) => ({
+          id,
+          values: flag.values,
+        }));
+
+        const tags = Object.entries(cloudTargets.tags).map(([id, tag]) => {
+          const tagFlags = Object.entries(tag.flags).map(
+            ([flagId, tagFlag]) => ({ id: flagId, values: tagFlag.values })
+          );
+          return { id, tagFlags };
+        });
+
+        return {
+          releases,
+          targets,
+          flags,
+          tags,
+        };
+      },
+    }),
+    cloudFirmware: t.field({
+      type: CloudFirmware,
+      args: {
+        params: t.arg({ type: CloudFirmwareParams, required: true }),
+      },
+      resolve: async (_, { params }, { cloudbuild }) => {
+        const jobStatus = await cloudbuild.queryJobStatus(params);
+        return {
+          status: jobStatus.status,
+          downloadUrl: jobStatus.artifacts[0].download_url,
+        };
+      },
+    }),
+  }),
+});
+
+builder.mutationType({
+  fields: (t) => ({
+    createCloudFirmware: t.field({
+      type: CloudFirmware,
+      args: {
+        params: t.arg({ type: CloudFirmwareParams, required: true }),
+      },
+      resolve: async (_, { params }, { cloudbuild }) => {
+        const jobStatus = await cloudbuild.createJob(params);
+        return {
+          status: jobStatus.status,
+          downloadUrl: jobStatus.artifacts[0].download_url,
+        };
+      },
+    }),
+  }),
+});
+
+builder.objectFields(CloudFirmware, (t) => ({
+  base64Data: t.string({
```

Comment:
Nice! This is what I meant :)

This way you don't "download" the data unless you request it